cart-service/src/cart/services/cart.service.ts

```diff
@@ -34,22 +46,58 @@ export class CartService {
     return this.createByUserId(userId);
   }
 
-  updateByUserId(userId: string, { items }: Cart): Cart {
-    const { id, ...rest } = this.findOrCreateByUserId(userId);
-
-    const updatedCart = {
-      id,
-      ...rest,
-      items: [ ...items ],
+  async updateByUserId(userId: string, item: LegacyCartItem): Promise<any> {
+    const { id, ...rest } = await this.findOrCreateByUserId(userId);
+    const cart = await this.findOrCreateByUserId(userId);
+    console.log('cart', cart);
+
+    let cartItem = cart.items.find(
+      (record) => record.productId === item.product.id,
+    );
+    if (cartItem) {
+      if (item.count === 0) {
+        await this.cartItemRepository.remove(cartItem);
+      }
+
+      cartItem.count = item.count;
+      await this.cartItemRepository.save(cartItem);
+    } else {
+      const product = new Product();
+      product.id = item.product.id;
+      product.title = item.product.title;
+      product.description = item.product.description;
+      product.price = item.product.price;
+      await this.productRepository.save(product);
+
+      cartItem = new CartItem();
+      cartItem.cartId = cart.id;
+      cartItem.productId = item.product.id;
+      cartItem.count = item.count;
+      cartItem.product = product;
+
+      await this.cartItemRepository.save(cartItem);
     }
 
-    this.userCarts[ userId ] = { ...updatedCart };
+    const updatedItems = await this.cartItemRepository.find({
+      where: { cartId: id },
+      relations: ['product'],
+    });
 
-    return { ...updatedCart };
+    return updatedItems;
   }
 
-  removeByUserId(userId): void {
-    this.userCarts[ userId ] = null;
-  }
+  async removeByUserId(userId): Promise<void> {
+    const cart = await this.cartRepository.findOne({ userId });
+    const cartItem = await this.cartItemRepository.find({ cartId: cart.id });
+    const productIDs = cartItem.map((record) => record.productId);
 
+    const pResponse = await this.productRepository.delete({
```

Comment:
I'm not sure that product should be deleted, when a cart item is deleted. 
For example, 'user1' wants to buy book 'Book1' and 'user2' wants to buy 'Book1'. When a cart for 'user1' is deleted, then the cart of 'user2' seems to be in invalid state.